js/src/features/project/components/ProjectRolesForm.tsx

```diff
@@ -0,0 +1,156 @@
+  }
+
+  function editRole(index: number) {
+    const fields = form.getFieldsValue()
+    edit({
+      roles: fields.roles[index],
+      slug: project.slug,
+    })
+  }
+
+  function removeRole(index: number) {
+    const fields = form.getFieldsValue()
+    remove({
+      roleId: fields.roles[index].id,
+      slug: project.slug,
+    })
+  }
+
+  function createRole(index: number) {
+    const fields = form.getFieldsValue()
+    create({
+      roles: fields.roles[index],
+      slug: project.slug,
+    })
+  }
+
+  return (
+    <Form
+      form={form}
+      labelCol={{span: 6}}
+      onFinishFailed={onFinishFailed}
+      initialValues={initialValues}
+    >
+      <Form.List name="roles">
+        {(fields, {add, remove}) => {
+          return (
+            <div>
+              {fields.map((field, index) => {
+                return (
+                  <React.Fragment key={index}>
+                    <Form.Item
+                      name={[field.name, "name"]}
+                      label={["Name"]}
+                      rules={[{required: true, message: "Please input Name!"}]}
+                    >
+                      <Input placeholder="last name"/>
+                    </Form.Item>
+                    <Form.Item
+                      name={[field.name, "permissions"]}
+                      label={["Permissions"]}
+
+                    >
+                      <Select mode="tags" style={{width: '100%'}} tokenSeparators={[',']}>
+                        {getSelectionChildren(index)}
+                      </Select>
+                    </Form.Item>
+
+                    {!!project.roles[index] ?
+                      <FormRoleControlButtons
+                        loading={loading}
+                        action={() => editRole(index)}
+                        remove={() => removeRole(index)}
+                      /> :
+                      <FormRoleControlButtons
+                        loading={loading}
+                        action={() => createRole(index)}
+                        remove={() => remove(index)}
+                      />}
+
+                    <div style={{borderBottom: "1px solid #f0f0f0", marginBottom: "24px"}}/>
+                  </React.Fragment>
+                )
+              })}
+              <Form.Item>
+                <Button
+                  type="dashed"
+                  onClick={() => add()}
+                  style={{width: "100%"}}
+                  loading={loading}
+                >
+                  <PlusOutlined/> Add
+                </Button>
+              </Form.Item>
+            </div>
+          )
+        }}
+      </Form.List>
+    </Form>
+  )
+}
+
+interface FormRoleControlButtons {
+  remove: () => void
+  action: () => void
+  loading: boolean
+}
+
+const FormRoleControlButtons: React.FC<FormRoleControlButtons> = ({remove, action, loading}) => {

```

Comment:
Why didn't you move this in separate file?

Reply:
And why your props and your component have the same names `FormRoleControlButtons`?